code-style.md

```diff
@@ -0,0 +1,120 @@
+# Code style guidelines
+
+Precise guidelines for code formatting are not yet available.
+The following guidelines are concerned with more high-level practices.
+
+We use the [Google Java style guide](https://google.github.io/styleguide/javaguide.html).
+We refine some of the guidelines here. Unless specified otherwise,
+these guidelines should be applied to Xtend, Java and Kotlin files.
+
+### Ordering of source files
+
+https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents
+
+> The order you choose for the members and initializers of your class can have a great effect on learnability. However, there's no single correct recipe for how to do it; different classes may order their contents in different ways.
+
+> What is important is that each class uses some logical order, which its maintainer could explain if asked. For example, new methods are not just habitually added to the end of the class, as that would yield "chronological by date added" ordering, which is not a logical ordering.
+
+Currently some of classes are sorted by visibility in very boldly delimited "sections".
+
+This type of sorting is bad practice, as a change in visibility requires the member to be moved around in the class, generally far away. Related members that don't have the same visibility may also be far away.
+
+A more usual practice in java is to
+- keep all fields at the very top (and sort them internally by visibility)
+- then declare constructors
+- then put all methods, in "some logical order". That may mean, a public method, then its three private implementation helpers, then another public method, etc.
+- put inner classes at the very end
+
+Keeping fields and constructors next to each other helps understanding the initialization logic of the object. Grouping fields at the top helps understanding at a glance what the object contains, and so, what is its lifetime in the program (is it a data class or identity based? does it contain references to big data structures (eg the AST)? can I copy this efficiently?).
+
+
+### Visibility
+
+- avoid public non-final fields
+- avoid public collection fields (even if final, they may be mutable)
+
+### Design for inheritance or prohibit it
+
+Make new classes final by default, and don't use protected members.
+The class should only be unsealed if a use case for inheritance shows up, which is in many cases never.
```

Comment:
There was a comment from Alexander some months ago that went something like, "If you make a class final, you patronize the programmer... Sometimes you just need to change one method, and then it would work perfectly... One time [someone] became so frustrated that he copied an entire class, pasted it into a new file, and called it [name of the original class]2." I think that this issue does not really apply to us since we use a different project structure where we do not have multiple separate projects that only depend on specific releases. As a result, the use of the `final` keyword for us would in practice become a fairly mild constraint -- basically just a suggestion and a signal of the original programmer's intent. Therefore it seems unnecessary for this issue to further delay this PR.